### DISCUSSION TITLE
Much needed features

### DISCUSSION THREAD

**Discussion by VictorZakharov** [Kilvoctu/aiyabot#157] [Ideas]
Been using the bot for a while, it's great, but a few things missing:
- "You can't use other people's [pencil icon]" <-- need to be able to. Does not make sense to block, anyone can view image details, and just type them in, but takes too long.
- When editing my own prompt with initial source (uploaded image or url), that initial source is not retained. So need to reupload and retype - inefficient workflow.
- Would be great if it was possible to use any generated image (or otherwise uploaded to the same discord channel) as a prompt image for the next generation, which will either use init image as an upload or an image URL internally. Then set the strength and experiment with prompt/output.

**Comment by Kilvoctu** [comment]
Heya Victor, yes I remember you. You wanted the moderation tools before.

> Would be great if it was possible to use any generated image (or otherwise uploaded to the same discord channel) as a prompt image for the next generation, which will either use init image as an upload or an image URL internally  

I'm not sure what you mean by this. You can right-click any image in Discord, then select Copy Image or Copy Link, then use **init_img/init_url** respectively to use it for /draw

For the other two things, I added it to the [projects list](https://github.com/Kilvoctu/aiyabot/projects?query=is%3Aopen), [here](https://github.com/users/Kilvoctu/projects/12/views/1?pane=issue&itemId=23951794) and [here](https://github.com/users/Kilvoctu/projects/15/views/1?pane=issue&itemId=23951975). I think it's not too hard to make those changes, but they'll probably have to wait for a while.

**Reply by VictorZakharov** [reply]
@Kilvoctu Yep, that's me. My nephew has been enjoying SD with aiya.

Regarding right-clicking an image in Discord, somehow I thought it includes a link to a thumbnail, which is of worse quality. Cause it opens very small in the browser. Turns out - not the case. When saved, the image is of actual size as uploaded.

Still, would be great to have a button number 5 which would toggle prompt edit with img2img mode. Currently you cannot include an image when editing a prompt. It could just add another field called image URL, prepopulated from that prompt I selected for edit. Or if you can add a toggle from the regular edit to include img2img, then no need for button number 5 in the main row underneath the image. Need to be able to set strength also in the same UI.

**Reply by Kilvoctu** [reply]
Yes, when AIYA uploads the generated image, it's a file attachment of the actual image generated by Web UI, not a link. That's partly why it takes so long for Discord to process it or whatever it's doing. There is no quality loss.

I won't need to add an extra button for img2img; it can simply be added to 🖋️. The modal has room for five fields, so there is one more space to use for the image url.
The "extended edit" in the modal already supports strength. It gets added there when **init_img/init_url** is used. Otherwise, it can be typed in manually once I add the image url to the modal.

**Reply by VictorZakharov** [reply]
@Kilvoctu There is an edge case -- if source image for edit contains init_img or url, then which url should it include in the edit? I think it should be init_img, but there should be an option to toggle url of the current image. On the same note, if init_img was not used, then editing the prompt should not include img2img by default, but it should be possible to toggle, at which point `strength` field will also show up. Otherwise `strength` should be hidden.

I am not sure how to implement this. Will you need a separate dialog? Leaving the details up to you.

**Reply by Kilvoctu** [reply]
First request done with https://github.com/Kilvoctu/aiyabot/commit/4b893bda736fc153add0bb9f60a9baccec4faece; it was pretty simple to do.  

Discord Modals are incredibly basic. It's up to five simple text fields and no other features; no toggles, dropdowns, buttons, etc.. Hopefully Discord can improve this UI framework at some point. 
Anyway, I'll think of something. If anything is not too clear, I do try to document usage in the [wiki](https://github.com/Kilvoctu/aiyabot/wiki).

**Reply by VictorZakharov** [reply]
@Kilvoctu Thanks, is it now configurable in config.toml ? Or need to mess with py files?

**Reply by Kilvoctu** [reply]
It's in the config file under `restrict_buttons` I think. Sorry I forgot to update the wiki page for it.

**Reply by VictorZakharov** [reply]
@Kilvoctu You can also update `config.toml` with defaults. Then I know where/how to change it. Like you did with `prompt_ban_list`.
But I cannot test it anyway, cause I'm currently alone on my discord server.

**Reply by Kilvoctu** [reply]
AIYA already does that. On every launch, she runs a check for discrepancies in the config, then if one is found she will add (or remove) keys using `default_config` as template, then port over any user settings. The console log will also note what's happening (though it's easy to miss it). This is the code that does it
https://github.com/Kilvoctu/aiyabot/blob/4b893bda736fc153add0bb9f60a9baccec4faece/core/settings.py#L358-L375

`default_config` is defined at the top of settings.py with all the defaults (bot hosts should not need to touch this)

**Reply by VictorZakharov** [reply]
@Kilvoctu I see. Can you explain how these work?
```
batch = "1,1"
max_batch = "8,1"
```
I need to set max count to 8, otherwise need to set the settings every time after startup.
I tried max_batch 1,8 and also 8,1, no effect, max size is 1.

**Reply by VictorZakharov** [reply]
@Kilvoctu I think figured it out, I tried to set batch 8, which assumes batch size 8, instead Aiya should figure out how to best serve that request from the allowed settings. For example, both 8,1 and 1,8 satisfy the 8 requirement.

Also I noticed the new version uploads each image individually to discord, and only shows image controls on the last image. This is inconvenient, I much prefer the old version where it grouped them together and gave nice small thumbnails. I could also navigate with keyboard arrow keys <-- all this functionality was somehow lost. Would be great if you could bring it back, maybe as an option/configurable.

Old/desired behavior:
![image](https://user-images.githubusercontent.com/8978766/228388888-be732a57-a39f-490c-9f6a-70b3980e37ef.png)

**Reply by Kilvoctu** [reply]
For **batch** it is explained in the wiki here https://github.com/Kilvoctu/aiyabot/wiki/draw-command#batch. I'll duplicate this information in other wiki places that are relevant.

In regards to your last message, I've made a statement here https://github.com/Kilvoctu/aiyabot/discussions/153 (I should probably pin it).

**Reply by VictorZakharov** [reply]
@Kilvoctu So Aiya already does this. I must have found a bug then. I swear I tried both 1,8 and 8,1 then tried batch 8, but it said max count was only 1. But now my 8,1 works fine with batch 8, so not sure what happened and why it's working now.

> Automatic - This is a simpler way to set the batch and is generally recommended. For automatic, you only need to input one number: your desired number of images. AIYA will then try her best to generate that many pictures using the highest batch size. Example:

I commented on #153.
